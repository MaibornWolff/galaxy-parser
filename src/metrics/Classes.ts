import { QueryBuilder } from "../queries/QueryBuilder";
import { grammars } from "../helper/Grammars";
import { ExpressionMetricMapping } from "../app";
import { TreeParser } from "../helper/TreeParser";

export class Classes implements Metric {
    private classesStatementsSuperSet = [];
    private treeParser: TreeParser;

    constructor(allNodeTypes: ExpressionMetricMapping[], treeParser: TreeParser) {
        this.treeParser = treeParser;
        allNodeTypes.forEach((expressionMapping) => {
            if (
                expressionMapping.metrics.includes(this.getName()) &&
                expressionMapping.type === "statement"
            ) {
                const { expression } = expressionMapping;
                this.classesStatementsSuperSet.push("(" + expression + ") @" + expression);
            }
        });
    }

    calculate(parseFile: ParseFile): MetricResult {
        const tree = this.treeParser.getParseTree(parseFile);

        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements(this.classesStatementsSuperSet);

        const query = queryBuilder.build();
        const matches = query.matches(tree.rootNode);

        console.log(this.getName() + " - " + matches.length);

        return {
            metricName: this.getName(),
            metricValue: matches.length,
        };
    }

    getName(): string {
        return "classes";
    }
}
