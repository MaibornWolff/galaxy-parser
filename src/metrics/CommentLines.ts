import { QueryBuilder } from "../queries/QueryBuilder";
import fs from "fs";
import Parser from "tree-sitter";
import { grammars } from "../grammars";
import {ExpressionMetricMapping} from "../app";

export class CommentLines implements Metric {
    private commentStatementsSuperSet = [];

    constructor(allNodeTypes: ExpressionMetricMapping[]) {
        allNodeTypes.forEach((expressionMapping) => {
            if (expressionMapping.metrics.includes(this.getName()) && expressionMapping.type === "statement") {
                const { expression } = expressionMapping
                this.commentStatementsSuperSet.push("("+expression+") @" + expression)
            }
        })
    }

    calculate(parseFile: ParseFile): MetricResult {
        const treeSitterLanguage = grammars.get(parseFile.language);

        const parser = new Parser();
        parser.setLanguage(treeSitterLanguage);

        const sourceCode = fs.readFileSync(parseFile.filePath).toString();
        const tree = parser.parse(sourceCode);

        const queryBuilder = new QueryBuilder(treeSitterLanguage, tree);
        queryBuilder.setStatements(this.commentStatementsSuperSet);

        const query = queryBuilder.build();
        const matches = query.matches(tree.rootNode);

        const commentLines = matches.reduce((accumulator, match) => {
            const captureNode = match.captures[0].node;
            return accumulator + captureNode.endPosition.row - captureNode.startPosition.row + 1;
        }, 0);

        console.log(this.getName() + " - " + commentLines);

        return {
            metricName: this.getName(),
            metricValue: matches.length,
        };
    }

    getName(): string {
        return "comment_lines"
    }
}
