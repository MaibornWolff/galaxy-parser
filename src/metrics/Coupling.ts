import { QueryBuilder } from "../queries/QueryBuilder";
import { grammars } from "../helper/Grammars";
import { ExpressionMetricMapping } from "../app";
import { formatCaptures } from "../helper/Helper";
import { TreeParser } from "../helper/TreeParser";

interface Package {
    namespace: string;
    class: string;
    source: string;
}

interface Usage {
    usedNamespace: string;
    alias?: string;
    source: string;
}

export class Coupling implements CouplingMetric {
    private namespaceStatementsSuperSet = [
        `
        (
            (namespace_definition
                name: (namespace_name) @namespace_definition_name
            )
            (class_declaration
                (name) @class_name
            )+
        )
    `,
    ];
    private namespaceUseStatementsSuperSets = [
        [
            `
            (namespace_use_declaration
                (namespace_name) @namespace_name
                (namespace_use_group
                    (namespace_use_group_clause
                        (namespace_name) @namespace_use_item_name
                    )
                )
            )
        `,
        ],
        [
            `
            (namespace_use_clause
                (qualified_name) @namespace_use
                (namespace_aliasing_clause (name) @namespace_use_alias)?
            )+
        `,
        ],
    ];

    private treeParser: TreeParser;

    constructor(allNodeTypes: ExpressionMetricMapping[], treeParser: TreeParser) {
        this.treeParser = treeParser;
        allNodeTypes.forEach((expressionMapping) => {
            if (
                expressionMapping.metrics.includes(this.getName()) &&
                expressionMapping.type === "statement"
            ) {
                const { expression } = expressionMapping;
                this.namespaceStatementsSuperSet.push("(" + expression + ") @" + expression);
            }
        });
    }

    calculate(parseFiles: ParseFile[]): CouplingMetricResult {
        let packages: Map<string, Package> = new Map();
        let usages: Usage[] = [];

        for (const parseFile of parseFiles) {
            packages = new Map([...packages, ...this.getPackages(parseFile)]);

            usages = usages.concat(this.getGroupedUsages(parseFile));
            usages = usages.concat(this.getSimpleUsages(parseFile));
        }

        console.log("\n\n");
        console.log("packages", packages, "\n\n");
        console.log("usages", usages);

        const couplingResults = this.getCoupledFilesData(packages, usages);

        console.log("\n\n", couplingResults);

        return {
            metricName: this.getName(),
            metricValue: couplingResults,
        };
    }

    /**
     * TODO scan interface, abstract class, and trait declarations as well (not only classes)
     */
    private getPackages(parseFile: ParseFile) {
        const packages: Map<string, { namespace: string; class: string; source: string }> =
            new Map();

        const tree = this.treeParser.getParseTree(parseFile);

        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements(this.namespaceStatementsSuperSet);

        const query = queryBuilder.build();
        const captures = query.captures(tree.rootNode);
        const textCaptures = formatCaptures(tree, captures);

        console.log(textCaptures);

        for (let index = 0; index < textCaptures.length; index += 1) {
            const namespaceName = textCaptures[index].text;

            let hasClassDeclaration = textCaptures[index + 1]?.name === "class_name";
            let classDeclarationIndex = index;

            while (hasClassDeclaration) {
                const className = textCaptures[classDeclarationIndex + 1].text;
                packages.set(namespaceName + "\\" + className, {
                    namespace: namespaceName,
                    class: className,
                    source: parseFile.filePath,
                });

                hasClassDeclaration =
                    textCaptures[classDeclarationIndex + 2]?.name === "class_name";
                classDeclarationIndex++;
                index++;
            }
        }

        return packages;
    }

    private getGroupedUsages(parseFile: ParseFile) {
        const tree = this.treeParser.getParseTree(parseFile);

        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements(this.namespaceUseStatementsSuperSets[0]);

        const usagesQuery = queryBuilder.build();
        const usagesCaptures = usagesQuery.captures(tree.rootNode);
        const usagesTextCaptures = formatCaptures(tree, usagesCaptures);

        console.log(usagesTextCaptures);

        const usagesOfFile: {
            usedNamespace: string;
            usedSource: string;
            alias?: string;
            source: string;
        }[] = [];

        for (let index = 0; index < usagesTextCaptures.length; index++) {
            const namespaceName = usagesTextCaptures[index].text;

            let hasUseGroupItem = usagesTextCaptures[index + 1]?.name === "namespace_use_item_name";
            let groupItemIndex = index;

            while (hasUseGroupItem) {
                const nextUseItem = usagesTextCaptures[groupItemIndex + 1];

                usagesOfFile.push({
                    usedNamespace: namespaceName + "\\" + nextUseItem.text,
                    usedSource: parseFile.filePath,
                    alias: "",
                    source: parseFile.filePath,
                });

                hasUseGroupItem =
                    usagesTextCaptures[groupItemIndex + 2]?.name === "namespace_use_item_name";
                groupItemIndex++;
                index++;
            }
        }

        return usagesOfFile;
    }

    private getSimpleUsages(parseFile: ParseFile) {
        const tree = this.treeParser.getParseTree(parseFile);

        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements(this.namespaceUseStatementsSuperSets[1]);

        const usagesQuery = queryBuilder.build();
        const usagesCaptures = usagesQuery.captures(tree.rootNode);
        const usagesTextCaptures = formatCaptures(tree, usagesCaptures);

        console.log(usagesTextCaptures);

        const usagesOfFile: {
            usedNamespace: string;
            usedSource: string;
            alias?: string;
            source: string;
        }[] = [];
        for (let index = 0; index < usagesTextCaptures.length; index++) {
            if (usagesTextCaptures[index].name === "namespace_use_alias") {
                const namespaceAlias = usagesTextCaptures[index].text;
                usagesOfFile[index - 1].alias = namespaceAlias;

                continue;
            }

            const namespaceName = usagesTextCaptures[index].text;
            usagesOfFile.push({
                usedNamespace: namespaceName,
                usedSource: parseFile.filePath,
                alias: "",
                source: parseFile.filePath,
            });
        }

        return usagesOfFile;
    }

    private getCoupledFilesData(packages: Map<string, Package>, usages: Usage[]) {
        return usages.flatMap((usage) => {
            if (packages.has(usage.usedNamespace)) {
                const fromData = [...packages.values()].filter((value) => {
                    if (usage.source === value.source) {
                        return value;
                    }
                });
                if (fromData.length > 0) {
                    const firstFromNamespace = fromData[0];
                    return [
                        {
                            fromNamespace:
                                firstFromNamespace.namespace + "\\" + firstFromNamespace.class,
                            toNamespace: usage.usedNamespace,
                            fromSource: usage.source,
                            toSource: packages.get(usage.usedNamespace).source,
                        },
                    ];
                }
            }
            return [];
        });
    }

    getName(): string {
        return "coupling";
    }
}
