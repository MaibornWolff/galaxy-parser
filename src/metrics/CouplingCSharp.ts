import { QueryBuilder } from "../queries/QueryBuilder";
import {Tree} from "tree-sitter";
import { grammars } from "../grammars";
import {ExpressionMetricMapping} from "../app";
import {formatCaptures} from "../helper/Helper";
import {TreeParser} from "../helper/TreeParser";

interface Package {
    namespace: string, class: string, source: string
}

interface Usage {
    usedNamespace: string, alias?: string, source: string
}

const NAMESPACE_DELIMITER = "."

export class CouplingCSharp implements CouplingMetric {
    private namespaceStatementsSuperSet = [`
            (namespace_declaration
                name: (_) @namespace_definition_name
                body: (_ (class_declaration name: (_) @class_name))
            )
    `];
    private namespaceUseStatementsSuperSets = [
        [`
            (using_directive
                (qualified_name) @namespace_use
            )
        `],
        [`
            (object_creation_expression
                type: (_) @object_class_name
            )
            (member_access_expression) @member_access_expression
        `]
    ];
    private treeParser: TreeParser

    constructor(allNodeTypes: ExpressionMetricMapping[], treeParser: TreeParser) {
        this.treeParser = treeParser;

        allNodeTypes.forEach((expressionMapping) => {
            if (expressionMapping.metrics.includes(this.getName()) && expressionMapping.type === "statement") {
                const { expression } = expressionMapping
                this.namespaceStatementsSuperSet.push("("+expression+") @" + expression)
            }
        })
    }

    calculate(parseFiles: ParseFile[]): CouplingMetricResult {
        let packages: Map<string, Package> = new Map();
        let usages: Usage[] = [];

        for (const parseFile of parseFiles) {
            packages = new Map([...packages, ...this.getPackages(parseFile)])
        }

        for (const parseFile of parseFiles) {
            const tree = this.treeParser.getParseTree(parseFile)

            usages = usages.concat(this.getSimpleUsages(parseFile, tree))
            usages = this.getEnrichedUsageCandidates(packages, usages, parseFile, tree)
        }

        console.log("\n\n")
        console.log("packages", packages, "\n\n")
        console.log("usages", usages)

        const couplingResults = this.getCoupledFilesData(packages, usages)

        console.log("\n\n", couplingResults)

        return {
            metricName: this.getName(),
            metricValue: couplingResults,
        };
    }

    /**
     * TODO scan interface, abstract class, and trait declarations as well (not only classes)
     */
    private getPackages(parseFile: ParseFile) {
        const packages: Map<string, {namespace: string, class: string, source: string}> = new Map();

        const tree = this.treeParser.getParseTree(parseFile)

        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements(this.namespaceStatementsSuperSet);

        const query = queryBuilder.build();
        const captures = query.captures(tree.rootNode);
        const textCaptures = formatCaptures(tree, captures);

        console.log(textCaptures);

        for (let index = 0; index < textCaptures.length; index+=1) {
            const namespaceName = textCaptures[index].text

            let hasClassDeclaration = textCaptures[index + 1]?.name === "class_name"
            let classDeclarationIndex = index;

            while(hasClassDeclaration) {
                const className = textCaptures[classDeclarationIndex + 1].text
                packages.set(namespaceName + NAMESPACE_DELIMITER + className, {namespace: namespaceName, class: className, source: parseFile.filePath})

                hasClassDeclaration = textCaptures[classDeclarationIndex + 2]?.name === "class_name"
                classDeclarationIndex++
                index++
            }
        }

        return packages
    }

    private getEnrichedUsageCandidates(packages: Map<string, Package>, usages: Usage[], parseFile: ParseFile, tree: Tree) {
        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements(this.namespaceUseStatementsSuperSets[1]);

        const usagesQuery = queryBuilder.build();
        const usagesCaptures = usagesQuery.captures(tree.rootNode);

        const objectCreations = usagesCaptures.filter((capture) => {
            return capture.name === "object_class_name";
        });

        const rawObjectCreations = formatCaptures(tree, objectCreations);
        const objectsCreated = new Set();
        for (const objectCreation of rawObjectCreations) {
            objectsCreated.add(objectCreation.text)
        }

        const memberAccesses = usagesCaptures.filter((capture) => {
            return capture.name === "member_access_expression";
        });

        const rawMemberAccesses = formatCaptures(tree, memberAccesses);

        const accessedClassNames = new Set();
        const regexValidClassNames = new RegExp(/^[A-Z][A-Za-z0-9_]*\.[A-Z][A-Za-z0-9_]*/);
        for (const member of rawMemberAccesses) {
            if (!regexValidClassNames.test(member.text)) {
                continue;
            }
            if (member.text.charAt(0) === member.text.charAt(0).toUpperCase()) {
                const className = member.text.substring(0, member.text.lastIndexOf("."))
                accessedClassNames.add(className)
            }
        }

        //console.log("\n\n")
        //console.log("CREATIONS", objectsCreated)
        //console.log("Accesseds", accessedClassNames)
        //console.log("length of usagse: ", usages.length)
        //console.log("\n\n")

        const usageCandidates: Map<string, Usage> = new Map()

        for (const usage of usages) {
            const originalUsage: Usage = {...usage}
            usageCandidates.set(originalUsage.usedNamespace, originalUsage)

            for (const objectCreation of objectsCreated) {
                const candidate: Usage = {...usage}
                candidate.usedNamespace += NAMESPACE_DELIMITER + objectCreation
                if (packages.has(candidate.usedNamespace)) {
                    usageCandidates.set(candidate.usedNamespace, candidate)
                }
            }
            for (const accessedClassName of accessedClassNames) {
                const candidate: Usage = {...usage}
                candidate.usedNamespace += NAMESPACE_DELIMITER + accessedClassName
                if (packages.has(candidate.usedNamespace)) {
                    usageCandidates.set(candidate.usedNamespace, candidate)
                }
            }
        }
        //console.log("usage candidates: ", usageCandidates)
        //console.log("\n\n")

        return [...usageCandidates.values()]
    }

    private getSimpleUsages(parseFile: ParseFile, tree: Tree) {
        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements(this.namespaceUseStatementsSuperSets[0]);

        const usagesQuery = queryBuilder.build();
        const usagesCaptures = usagesQuery.captures(tree.rootNode);
        const usagesTextCaptures = formatCaptures(tree, usagesCaptures);

        console.log(usagesTextCaptures);

        const usagesOfFile: {usedNamespace: string, usedSource: string, alias?: string, source: string}[] = [];
        for (let index = 0; index < usagesTextCaptures.length; index++) {
            if (usagesTextCaptures[index].name === "namespace_use_alias") {
                const namespaceAlias = usagesTextCaptures[index].text
                usagesOfFile[index - 1].alias = namespaceAlias

                continue;
            }

            const namespaceName = usagesTextCaptures[index].text
            usagesOfFile.push({usedNamespace: namespaceName, usedSource: parseFile.filePath, alias: "", source: parseFile.filePath})
        }

        return usagesOfFile
    }

    private getCoupledFilesData(packages: Map<string, Package>, usages: Usage[]) {
        return usages.flatMap((usage) => {
            if (packages.has(usage.usedNamespace)) {
                const fromData = [...packages.values()].filter((value) => {
                    if (usage.source === value.source) {
                        return value;
                    }
                });
                if (fromData.length > 0) {
                    const firstFromNamespace = fromData[0]
                    return [{
                        fromNamespace: firstFromNamespace.namespace + NAMESPACE_DELIMITER + firstFromNamespace.class,
                        toNamespace: usage.usedNamespace,
                        fromSource: usage.source,
                        toSource: packages.get(usage.usedNamespace).source
                    }]
                }
            }
            return [];
        })
    }

    getName(): string {
        return "coupling"
    }
}
