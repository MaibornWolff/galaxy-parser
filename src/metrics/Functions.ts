import { QueryBuilder } from "../queries/QueryBuilder";
import { grammars } from "../grammars";
import {ExpressionMetricMapping} from "../app";
import {TreeParser} from "../helper/TreeParser";

export class Functions implements Metric {
    private functionsStatementsSuperSet = [];
    private treeParser: TreeParser

    constructor(allNodeTypes: ExpressionMetricMapping[], treeParser: TreeParser) {
        this.treeParser = treeParser;

        allNodeTypes.forEach((expressionMapping) => {
            if (expressionMapping.metrics.includes(this.getName()) && expressionMapping.type === "statement") {
                const { expression } = expressionMapping
                this.functionsStatementsSuperSet.push("("+expression+") @" + expression)
            }
        })
    }

    calculate(parseFile: ParseFile): MetricResult {
        const tree = this.treeParser.getParseTree(parseFile);

        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements(this.functionsStatementsSuperSet);

        const query = queryBuilder.build();
        const matches = query.matches(tree.rootNode);

        console.log(this.getName() + " - " + matches.length);

        return {
            metricName: this.getName(),
            metricValue: matches.length,
        };
    }

    getName(): string {
        return "functions"
    }
}
