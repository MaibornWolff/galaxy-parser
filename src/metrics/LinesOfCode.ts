import { QueryBuilder } from "../queries/QueryBuilder";
import { grammars } from "../helper/Grammars";
import { TreeParser } from "../helper/TreeParser";
import { ExpressionMetricMapping } from "../helper/Model";

export class LinesOfCode implements Metric {
    private startRuleStatementsSuperSet = [];
    private treeParser: TreeParser;

    constructor(allNodeTypes: ExpressionMetricMapping[], treeParser: TreeParser) {
        this.treeParser = treeParser;
        allNodeTypes.forEach((expressionMapping) => {
            if (
                expressionMapping.metrics.includes(this.getName()) &&
                expressionMapping.type === "statement"
            ) {
                const { expression } = expressionMapping;
                this.startRuleStatementsSuperSet.push("(" + expression + ") @" + expression);
            }
        });
    }

    calculate(parseFile: ParseFile): MetricResult {
        const tree = this.treeParser.getParseTree(parseFile);

        const queryBuilder = new QueryBuilder(grammars.get(parseFile.language), tree);
        queryBuilder.setStatements(this.startRuleStatementsSuperSet);

        const query = queryBuilder.build();
        const matches = query.matches(tree.rootNode);

        const loc = matches.length > 0 ? matches[0].captures[0].node.endPosition.row : 0;
        console.log(this.getName() + " - " + loc);

        return {
            metricName: this.getName(),
            metricValue: matches.length,
        };
    }

    getName(): string {
        return "lines_of_code";
    }
}
