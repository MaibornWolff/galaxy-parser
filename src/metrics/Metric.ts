interface MetricResult {
    metricName: string;
    metricValue: number;
}

interface CouplingMetricValue {
    fromNamespace: string, toNamespace: string, fromSource: string, toSource: string
}

interface CouplingMetricResult {
    metricName: string;
    metricValue: CouplingMetricValue[];
}

interface Metric {
    calculate(parseFile: ParseFile): MetricResult;
    getName(): string;
}

interface CouplingMetric {
    calculate(files: ParseFile[]): CouplingMetricResult;
    getName(): string;
}

interface ParseFile {
    language: string;
    filePath: string;
}
