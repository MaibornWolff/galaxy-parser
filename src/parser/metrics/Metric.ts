/**
 * Interface for carrying the result of a metric calculation.
 */
export interface MetricResult {
    /**
     * Name of the metric.
     */
    metricName: string;

    /**
     * Value of the metric.
     */
    metricValue: number;
}

export interface Relationship {
    fromNamespace: string;
    toNamespace: string;
    fromSource: string;
    toSource: string;
    fromClassName: string;
    toClassName: string;
    usageType: string | "usage" | "extends" | "implements";
}

export interface CouplingMetrics {
    outgoing_dependencies: number;
    incoming_dependencies: number;
    coupling_between_objects: number;
    instability: number;
}

export interface CouplingResult {
    relationships: Relationship[];
    metrics: Map<string, CouplingMetrics>;
}

/**
 * Interface for calculating a concrete metric.
 */
export interface Metric {
    /**
     * Calculates the metric value for the supplied file.
     * @param parseFile Source code file for which the metric value should be calculated.
     * @return A MetricResult containing the calculated metric value.
     */
    calculate(parseFile: ParseFile): MetricResult;

    /**
     * Returns the name of this metric.
     * @return The name of this metric.
     */
    getName(): string;
}

export interface CouplingMetric {
    calculate(files: ParseFile[]): CouplingResult;
    getName(): string;
}

/**
 * Represents a file to be parsed, including its path and the file extension.
 */
export interface ParseFile {
    /**
     * File extension of the file.
     */
    fileExtension: string;

    /**
     * Path to the file.
     */
    filePath: string;
}
