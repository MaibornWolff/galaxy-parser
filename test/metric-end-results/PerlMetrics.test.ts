import { expectFileMetric, getCouplingMetrics, parseAllFileMetrics } from "./TestHelper";
import { FileMetric, FileMetricResults } from "../../src/parser/metrics/Metric";

describe("Perl metrics tests", () => {
    const path = "./resources/perl/";

    let results: Map<string, FileMetricResults>;

    const testFileMetric = (inputPath: string, metric: FileMetric, expected: number) =>
        expectFileMetric(results, inputPath, metric, expected);

    beforeAll(async () => {
        results = await parseAllFileMetrics(path);
    });

    describe("parsing complexity metric", () => {
        it("should count compound statements correctly", () => {
            // count: if, elsif, unless, when, while, until, for, foreach
            // as function: BEGIN, UNITCHECK, CHECK, INIT, END
            // not: else, given, continue, {}
            testFileMetric(path + "compound-statements.pl", FileMetric.complexity, 28);
        });

        it("should NOT count dump", () => {
            // not: dump, CORE::dump
            testFileMetric(path + "dump.pl", FileMetric.complexity, 0);
        });

        it("should NOT count exit", () => {
            // not: exit
            testFileMetric(path + "exit.pl", FileMetric.complexity, 0);
        });

        it("should NOT count goto", () => {
            // (count: sub)
            // not: goto LABEL, goto EXPR, goto &NAME
            testFileMetric(path + "goto.pl", FileMetric.complexity, 3);
        });

        it("should NOT count loop control commands", () => {
            // (count: while, if)
            // not: next, last, redo
            testFileMetric(path + "loop-control.pl", FileMetric.complexity, 5);
        });

        it("should count operators correctly", () => {
            // count: and, &&, &&=, or, ||, ||=, //, //=, xor, ?:
            // not: everything else
            testFileMetric(path + "operators.pl", FileMetric.complexity, 22);
        });

        it("should count statement modifiers correctly", () => {
            // count: if, unless, while, until, for, foreach, when
            // not: given
            testFileMetric(path + "statement-modifiers.pl", FileMetric.complexity, 7);
        });

        it("should count functions correctly", () => {
            testFileMetric(path + "sub.pl", FileMetric.complexity, 8);
            testFileMetric(path + "sub-anonymous.pl", FileMetric.complexity, 6);
            testFileMetric(path + "sub-autoload.pl", FileMetric.complexity, 1);
            testFileMetric(path + "classes.pl", FileMetric.complexity, 4);
            testFileMetric(path + "classes-using-packages.pl", FileMetric.complexity, 4);
            testFileMetric(path + "defer-blocks.pl", FileMetric.complexity, 1);
        });

        it("should count switch statements correctly", () => {
            // count: when, for, foreach
            // not: given, default, continue, break
            testFileMetric(path + "switch-statements.pl", FileMetric.complexity, 12);
        });

        it("should count try-catch correctly", () => {
            // count: or, catch
            // not: eval, die, do, try
            testFileMetric(path + "try-catch.pl", FileMetric.complexity, 2);
        });
    });

    describe("parsing functions metric", () => {
        it("should count subroutines", () => {
            // count: sub definition
            // not: sub declaration
            testFileMetric(path + "sub.pl", FileMetric.functions, 8);
        });

        it("should count anonymous subroutines", () => {
            // count: sub
            testFileMetric(path + "sub-anonymous.pl", FileMetric.functions, 6);
        });

        it("should count AUTOLOAD", () => {
            // count: AUTOLOAD
            testFileMetric(path + "sub-autoload.pl", FileMetric.functions, 1);
        });

        it("should count class ADJUST and methods", () => {
            // count: ADJUST, method, sub, sub DESTROY
            testFileMetric(path + "classes.pl", FileMetric.functions, 4);
            testFileMetric(path + "classes-using-packages.pl", FileMetric.functions, 4);
        });

        it("should count compound statements", () => {
            // count: BEGIN, UNITCHECK, CHECK, INIT, END
            // not: if, else, elsif, unless, given, when, while, until, for, foreach, continue, {}
            testFileMetric(path + "compound-statements.pl", FileMetric.functions, 5);
        });

        it("should count defer blocks", () => {
            // count: defer
            testFileMetric(path + "defer-blocks.pl", FileMetric.functions, 1);
        });
    });

    describe("parsing classes metric", () => {
        it("should count classes correctly", () => {
            // count: class definition
            // not: class declaration
            testFileMetric(path + "classes.pl", FileMetric.classes, 2);
        });

        it("should count classes using packages correctly", () => {
            testFileMetric(path + "classes-using-packages.pl", FileMetric.classes, 2);
        });

        it("should NOT count normal packages", () => {
            testFileMetric(path + "modules/Module.pm", FileMetric.classes, 0);
        });
    });

    describe("parsing lines_of_code metric", () => {
        it("should count comments correctly", () => {
            testFileMetric(path + "comments.pl", FileMetric.linesOfCode, 11);
        });

        it("should count empty files correctly", () => {
            testFileMetric(path + "empty.pl", FileMetric.linesOfCode, 1);
        });

        it("should count one line break correctly", () => {
            testFileMetric(path + "line-break.pl", FileMetric.linesOfCode, 2);
        });

        it("should count one line correctly", () => {
            testFileMetric(path + "one-line.pl", FileMetric.linesOfCode, 1);
        });

        it("should count pods correctly", () => {
            testFileMetric(path + "pod.pl", FileMetric.linesOfCode, 27);
        });

        it("should count __DATA__ correctly", () => {
            testFileMetric(path + "special-literals-data.pl", FileMetric.linesOfCode, 12);
        });

        it("should count __END__ correctly", () => {
            testFileMetric(path + "special-literals.pl", FileMetric.linesOfCode, 20);
        });

        it("should count weird lines correctly", () => {
            testFileMetric(path + "weird-lines.pl", FileMetric.linesOfCode, 15);
        });
    });

    describe("parsing comment_lines metric", () => {
        describe("parsing comment_lines metric", () => {
            it("should count comments correctly", () => {
                // count: any line with #
                // not: # in strings
                testFileMetric(path + "comments.pl", FileMetric.commentLines, 5);
            });

            it("should count pods", () => {
                testFileMetric(path + "pod.pl", FileMetric.commentLines, 14);
            });

            it("should NOT count __DATA__", () => {
                testFileMetric(path + "special-literals-data.pl", FileMetric.commentLines, 2);
            });

            it("should count __END__", () => {
                // count: lines including and after __END__
                testFileMetric(path + "special-literals.pl", FileMetric.commentLines, 4);
            });
        });
    });

    describe("parsing real_lines_of_code metric", () => {
        it("should NOT count comments", () => {
            testFileMetric(path + "comments.pl", FileMetric.linesOfCode, 4);
        });

        it("should NOT count empty files", () => {
            testFileMetric(path + "empty.pl", FileMetric.linesOfCode, 0);
        });

        it("should NOT count one line break", () => {
            testFileMetric(path + "line-break.pl", FileMetric.linesOfCode, 0);
        });

        it("should count one line correctly", () => {
            testFileMetric(path + "line-break.pl", FileMetric.linesOfCode, 1);
        });

        it("should NOT count pods", () => {
            testFileMetric(path + "pod.pl", FileMetric.linesOfCode, 8);
        });

        it("should count __DATA__", () => {
            testFileMetric(path + "special-literals-data.pl", FileMetric.linesOfCode, 4);
        });

        it("should NOT count after __END__", () => {
            // count: line with __END__
            // not: lines after __END__
            testFileMetric(path + "special-literals.pl", FileMetric.linesOfCode, 12);
        });

        it("should count weird lines correctly", () => {
            testFileMetric(path + "weird-lines.pl", FileMetric.linesOfCode, 10);
        });
    });

    // describe("parsing max_nesting_level metric", () => {}); ???

    describe("parsing coupling metric", () => {
        it("should calculate correctly", async () => {
            const couplingResult = await getCouplingMetrics(path);
            expect(couplingResult).toMatchSnapshot();
        });
    });
});
