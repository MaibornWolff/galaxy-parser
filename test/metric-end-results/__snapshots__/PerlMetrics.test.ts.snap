// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Perl metrics tests parsing coupling metric should calculate correctly 1`] = `
{
  "metrics": Map {},
  "relationships": [],
}
`;
